### DISCUSSION TITLE
Lifecycle and retaining stores in memory

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [badoo/MVIKotlin#266] [Q&A] [answered]
So I'll try and properly articulate my confusion. Basically, Decompose via routers handles which child/ configuration is currently on top of the stack so their lifecycle as well is understandable. We can pop the stack so the lifecycle of Decompose components is pretty clear. Now MVIKotlin, stores can be preserved entirely or we can preserve an object, from all of the examples I've seen (from what I remember) we used `instanceKeeper.getStore(::calculatorStore)` meaning that we retain the whole store. So if we had 10 different screens we'd be retaining all 10 of these stores in memory? Is that a problem? I'd always just slap the `.getStore` and not think about it, but after diving a bit more into it, it seems that I should be more varied of how and what I retain? And how are even Stores "disposed" of, is it just waiting for the component to be popped from the stack and then the GC to clean it up (on Android)?

Also, how does the Lifecycle fit in this story? From what I've seen it's mostly used for binders, and if I follow a Single Activity Approach, I'd basically subscribe/ unsubscribe when my app is opening/ closing (and perhaps configuration changes)? Is there anything more to it that I am missing?

**Comment by arkivanov** [comment, marked as answer]
Hello @Nikola-Milovic and thanks for your questions!

1. Retained instances

As per [documentation](https://arkivanov.github.io/MVIKotlin/state_preservation.html), `InstanceKeeper` retains things (not only `Stores`) over their scope recreation. This means that objects are not retained in memory forever, but only until its scope (e.g. a Decompose component, or a Fragment) is alive or is being recreated due to Android configuration changes. So it is very similar to AndroidX `ViewModels`.

As of Decompose, when you pop a component, its retained instances are destroyed (`InstanceKeeper.Instance.onDestroy` method is caled). If there is a stack of components and you rotate the screen, only the currently active component is recreated and its `InstanceKeeper` instances are retained. All components in the back stack are destroyed with all their retained instances. This matches the behaviour of AndroidX Fragments.

2. How does the Lifecycle fit in this story?
 
In general, the `Lifecycle` is just a way to be notified when the `Lifecycle` state changes. So you can subscribe to its state changes and do whatever logic you need. The `Binder` is just on of the use cases, it allows to subscribe/unsubscribe streams based on the `Lifecycle` (e.g. keep a `View` subscribed to a `Store` only while the `Lifecycle` is `started`, so the `View` does not re-render while the app is in background). You can use the `Lifecycle` on your own as well. E.g. you may create Reaktive's `DisposableScope` and `dispose` it in `lifecycle.doOnDestroy {}`. Or you may create a `CoroutineScope` and `cancel` it in the same way.

If you follow the Single Activity approach, then you probably have something like Fragments or Decompose components, for each "screen". Each Fragment/Component has its own `Lifecycle`, which allows you to individually track which screens are active and which are not. When a Fragment/Component is recreated due to Android configuration changes, its `Lifecycle` goes as follows: `onStop` -> `onPause` -> `onDestroy` -> [new instance created] -> `onCreate` -> `onStart` -> `onResume`. Any objects retained via `InstanceKeeper` are not destroyed and just transferred to the new Fragment/Component.

**Reply by Nikola-Milovic** [reply]
Much simpler than I thought. Thank you, needed this clarification!
